### ISSUE TITLE
[XPU][XPUGraph][XCCL] XPUGraph replay with captured all_reduce can make later eager all_reduce output NaN

### ISSUE THREAD

**Issue by liu-shaojun** [pytorch/pytorch#189403]
### Describe the bug

We are debugging an intermittent NaN issue on Intel XPU. The original symptom was observed in vLLM, but we also have a standalone PyTorch/XPU reproducer with no vLLM dependency.

The standalone reproducer captures/replays XPU graphs containing distributed all_reduce, then runs later eager work with all_reduce. In this setup, the later eager all_reduce can produce NaN.

### Standalone reproducer

Attached:
- [xpu_graph_standalone_stress.py](https://github.com/user-attachments/files/29853187/xpu_graph_standalone_stress.py)
- [run_standalone_xpu_graph_ab.sh](https://github.com/user-attachments/files/29853222/run_standalone_xpu_graph_ab.sh)

Run command:

```bash
RESULT_ROOT=/tmp/standalone_ab_multigraph_h2048 \
ZE_AFFINITY_MASK=6,7 \
HIDDEN_SIZE=2048 \
ITERS=2000 \
CAPTURE_SIZES=1,2,3,4,5,6,7,8 \
POISON_CHURN=0 \
/path/to/run_standalone_xpu_graph_ab.sh
```

The script runs three modes:

1. graph_off
2. capture_skip_replay
3. capture_replay

Observed result from a fresh rerun:
```
graph_off: completed 2000/2000, no NaN
capture_skip_replay: completed 2000/2000, no NaN
capture_replay: NaN reproduced at iter1016.eager_layer0.post_collective, shape [25, 2048],
both ranks
```
Relevant log lines:
```
[rank0] {"event": "NONFINITE", "has_nan": true, "has_inf": false, "shape": [25, 2048],
"dtype": "torch.float16", "tag": "iter1016.eager_layer0.post_collective",
"overlap_graph_static": []}
[rank1] {"event": "NONFINITE", "has_nan": true, "has_inf": false, "shape": [25, 2048],
"dtype": "torch.float16", "tag": "iter1016.eager_layer0.post_collective",
"overlap_graph_static": []}
```
Additional isolation from previous runs:
```
capture_replay with POISON_CHURN=0:
NaN still reproduced at iter1016.eager_layer0.post_collective, shape [25, 2048], both
ranks.

capture_replay with sync_after_replay=1:
NaN still reproduced at iter1016.eager_layer0.post_collective, shape [25, 2048], both
ranks.

capture/replay with graph-side collective removed:
5000/5000, no NaN.

capture/replay with later eager collective removed:
5000/5000, no NaN.
```
The eager tensor is checked to be finite immediately before the eager all_reduce. The first non-finite value is observed immediately after the eager all_reduce.

So the minimal lower-level trigger appears to involve:

XPUGraph replay + captured XCCL/functional all_reduce + later eager all_reduce

### vLLM context

This issue was originally observed in vLLM LLM serving on Intel XPU. With XPU graph enabled, some requests occasionally produce all-! completions. The failure is intermittent: it may appear after a small or large number of requests, and a following request may become normal again.

We traced the all-! output to NaN logits. In vLLM, the first visible NaN location is not always stable, which suggests the symptom may depend on timing/allocation/runtime state rather than a single model kernel.

The standalone reproducer attached here is not intended to fully model vLLM. It isolates one lower-level pattern that looks related: XPUGraph replay with XCCL/functional collectives can make a later eager collective produce NaN.

So our current suspicion is broader than one captured collective op: there may be an interaction between XPUGraph replay and XCCL/eager collective state.

Do you have existing coverage for XPUGraph replay interleaved with XCCL and later eager collectives? Also, have you seen similar NaN/corruption issues around XPUGraph replay and XCCL before?

### Environment

torch: 2.11.0+xpu
device: Intel XPU
distributed backend: xccl

### Versions

curl -sL https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py | python
Collecting environment information...
PyTorch version: 2.11.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 4.3.4
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-1011-intel-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20250302
Intel GPU driver version:
* libze1:       1.28.2-1~24.04~ppa1
* intel-opencl-icd:     26.14.37833.4-1~24.04~ppa1
Intel GPU models onboard:
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
* Intel(R) Graphics [0xe223]
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-0e00-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [1] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-3400-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [2] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-4700-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [3] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-5a00-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [4] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-8600-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [5] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-ae00-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [6] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-c200-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [7] _XpuDeviceProperties(name='Intel(R) Graphics [0xe223]', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE223, uuid=868023e2-0000-0000-d600-000000000000, driver_version='1.15.37833+4', total_memory=32656MB, local_mem_size=128KB, max_compute_units=256, gpu_eu_count=256, gpu_subslice_count=32, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  64
On-line CPU(s) list:                     0-63
Vendor ID:                               GenuineIntel
BIOS Vendor ID:                          Intel(R) Corporation
Model name:                              INTEL(R) XEON(R) GOLD 6530
BIOS Model name:                         INTEL(R) XEON(R) GOLD 6530  CPU @ 2.1GHz
BIOS CPU family:                         179
CPU family:                              6
Model:                                   207
Thread(s) per core:                      1
Core(s) per socket:                      32
Socket(s):                               2
Stepping:                                2
CPU(s) scaling MHz:                      94%
CPU max MHz:                             4000.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               3 MiB (64 instances)
L1i cache:                               2 MiB (64 instances)
L2 cache:                                128 MiB (64 instances)
L3 cache:                                320 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-31
NUMA node1 CPU(s):                       32-63
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] dpcpp-cpp-rt==2025.3.2
[pip3] impi-rt==2021.17.2
[pip3] intel-cmplr-lib-rt==2025.3.2
[pip3] intel-cmplr-lib-ur==2025.3.2
[pip3] intel-cmplr-lic-rt==2025.3.2
[pip3] intel-opencl-rt==2025.3.2
[pip3] intel-openmp==2025.3.2
[pip3] intel-pti==0.16.0
[pip3] intel-sycl-rt==2025.3.2
[pip3] mkl==2025.3.1
[pip3] numpy==2.3.5
[pip3] oneccl==2021.17.2
[pip3] oneccl-devel==2021.17.2
[pip3] onemkl-license==2025.3.1
[pip3] onemkl-sycl-blas==2025.3.1
[pip3] onemkl-sycl-dft==2025.3.1
[pip3] onemkl-sycl-lapack==2025.3.1
[pip3] onemkl-sycl-rng==2025.3.1
[pip3] onemkl-sycl-sparse==2025.3.1
[pip3] tbb==2022.3.1
[pip3] tcmlib==1.4.1
[pip3] torch==2.11.0+xpu
[pip3] torchaudio==2.11.0+xpu
[pip3] torchvision==0.26.0+xpu
[pip3] triton-xpu==3.7.0
[pip3] umf==1.0.3
[conda] Could not collect

cc @gujinghui @EikanWang @fengyuan14 @guangyey

**Comment by guangyey** [comment]
cc @majing921201  Could you please help take a look.

**Comment by majing921201** [comment]
I will look into this issue. However, since we are currently upgrading the compiler and CCL, we'll evaluate it using the latest internal toolchain.
